### ISSUE TITLE
Create a db-utils package to abstract the nastier bits of the database code.

### ISSUE THREAD

**Issue by ChasingLogic** [ChasingLogic/cardinal#63]
Should all return promises to make our handlers easy to write.

**Comment by ChasingLogic** [comment]
From growlove:

Examples:
db.database.reset();
Drops all collections except for 'system.*' collections, which it empties instead.

db.collection.empty('magic');
Removes all objects from the magic collection.

I'm not sure if we really need db-utils for the entire project. I started working on it as something to go into test/support which we could import and then use in before functions to get the test database to a specific status, but I moved it out of there for now because I can see it possibly being useful elsewhere. shrug We'll see.

This is giving me ideas for how database/setup.js should work. I'm thinking a basic database schema which comes with a set of pre-defined functions, maybe something like this:

const schema = {
  collections: {
    magic: {
      indexes: {
        name: { unique: true }
      },
      setup(): {
        return async(function *() {
          // blah blah blah
          // db bullshit, using other information in this object to determine how to set it up
        })();
      }
    }
  }
};
Still definitely a concept-in-progress though. Any thoughts, @ChasingLogic?

**Comment by ChasingLogic** [comment]
If we want to go down that route (i.e. using predefined schemas) We should really look at an ORM library like Mongoose.(I generally despise ORMs but I could get over it if you think it's the way to go.) but it makes sense if our schemas only define how the collection is to be set up and not how they relate to go ahead and roll our own since that doesn't exist.

I really like the idea of abstracting the nastier database code away into promises and generators though and I think what you have here is perfect.

As a side note I've been using the async library should I switch to bluebird? Does it provide better/more functionality?

As a side side note, we should probably get in the habit of creating issues for whatever we are planning on working on / changing that way we know ahead of time because I now will have to rewrite large chunks of what I'm working on to fit your ( much much better ) model.

**Comment by ChasingLogic** [comment]
We could even make the db-utils return a "mock" db object for running tests with. i.e.

```javascript
const db = require('db-utils').newMockDB();

db.getCardByName('magic', 'lightning bolt');
```

Will still return an object. This makes our code easier to test. I still have concerns about how we test our db-utils / other database code and the work I've been doing on add-cards is relevant to this discussion.

https://github.com/ChasingLogic/cardinal/blob/master/database/add-cards.js
https://github.com/ChasingLogic/cardinal/blob/master/database/download-cards.js

Let's move this to Issue #63

**Comment by ChasingLogic** [comment]
From growlove:

I like the idea of a mock object. No need to even deal with the db itself
if we're just making sure the correct functions are being called to insert
the right objects.

On Wed, Nov 18, 2015, 7:55 AM Mathew Robinson <notifications@github.com>
wrote:

> We could even make the db-utils return a "mock" db object for running
> tests with. i.e.
>
> const db = require('db-utils').newMockDB();
>
> db.getCardByName('magic', 'lightning bolt');
>
> Will still return an object. This makes our code easier to test. I still
> have concerns about how we test our db-utils / other database code and the
> work I've been doing on add-cards is relevant to this discussion.
>
> https://github.com/ChasingLogic/cardinal/blob/master/database/add-cards.js
>
> https://github.com/ChasingLogic/cardinal/blob/master/database/download-cards.js
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/growlove/cardinal/commit/734d2b6f37e1b2bdfffb32a802751f7fe766c1d6#commitcomment-14470904>
> .


Mark A. Chandler, II <notifications@github.com>
8:43 AM (1 hour ago)

to growlove/cardi., me 
I meant less of a schema and more of an adapter for each card game's setup process. I'll post some more thoughts on it after work.

I've only been using Bluebird for its coroutine function. I think it does some other things similar to async but I haven't honestly dealt with either. If you like async I will rename the Bluebird object to 'co' to avoid conflict.

**Comment by ChasingLogic** [comment]
@growlove Async is no longer a dependency we'll just switch to bluebird it's a little different but we can get the some functionality out of bluebird.

Also as for the naming of the variable goes I think sticking with 

```javascript
const packageName = require('package-name');
```

Is going to be better and avoiding things like
```javascript
const somethingElse = require('package-name').someFunction;
```

Just for clarity. If you want to "alias" the function I think that's fine but should follow the 
```javascript
const someFunction = require('package-name').someFunction;
```

Form as a general rule. So I'd rename your import variable to coroutine in this case.

**Comment by growlove** [comment]
I'm fine with that rule. I'll switch things over to coroutine.

**Comment by ChasingLogic** [comment]
@growlove just an update on this I'm turning db-utils into a hapi plugin. It's more "hapi idiomatic" that way and can help us switch to TDD if you would like to. I still am not a personal fan of TDD but I can see some of the value.

**Comment by ChasingLogic** [comment]
Even with the move back to Go this would be useful to put in the database package. @growlove do you want to take this on?

**Comment by growlove** [comment]
Yes definitely, it'll be a good way to start building my go knowledge
